Part two updates (#1)
dfe-analytical-services/good-code-practice: 05-part_two.Rmd

# Advanced: Testing code{#testing} 

> From a coding point of view, testing covers the requirements spelled out in the **Verification** pillar of the QA framework. Note that testing does not overlap with the **Validation** QA pillar: it is concerned with code running properly rather than with the adequacy of analytical assumptions. Testing is a vast topic and it is beyond the scope of this guide to explore it in depth. Here we concentrate on broad principles corresponding to incrementally thorough testing. These ways of testing include:

Reviewer: ```suggestion
From a coding point of view, testing covers the requirements spelled out in the **Verification** pillar of the QA framework. Note that testing does not overlap with the **Validation** QA pillar: it is concerned with code running properly rather than with the adequacy of analytical assumptions. Testing is a vast topic and as such this guide will cover console tests and in-script testing in brief and focus on formalised testing procedures, such as unit testing.  These test the different components of "units" of code using tools such as testthat in R.
```

> - console tests 

Reviewer: Remove bullets as now covered in the intro

- console tests 
- in-script testing
- formalised testing procedures, such as unit testing (tests the different components or "units" of code using tools such as `testthat` in R)

## Console tests
Testing code within the R console is not recommended as it leaves no trace or records of what has been checked.

## In-script testing

> In-script testing is sufficient to facilitate QA in most cases, although tests should always be accompanied by brief comments signalling their presence and indicating what they are evaluating and what the expected output should be. For a quality assurer, running these tests ensures that the entire analytical workflow is reproducible. In R there are many ways to perform in-script checks, from `print()` statements, to `base` functions `stop()` and `stopfinot()`, to functions from dedicated packages such as `assertthat` or `testit`. Examples are shown below. 

Reviewer: Could we add something about Python and/or other languages here as well?  Even if it's just a simple statement such as: "In R there are........  Equivalent syntyax for Python and other languages is also readily available (any nice links??) but we've included some examples from R below"

Author: I'll need to do some research to address this one. I'll see if I can carve out some time later this week.

Author: I've upped the Python content now.

```

## Formalised testing

> Beyond in-script tests, an analyst may decide to take a more formal approach to testing if the project warrants it. Software development has dedicated testing professionals (called "testers") and there is plenty of literature about software testing on the internet. Much of this is transferable to code based analysis, but not all of it. There are times when it makes sense to make a distinction between a model (which is generally operated by a technical specialist) and an IT product (e.g. a website, which is used by a broader user base). Shiny apps sit somewhere between a model and an IT product so we'll include some specific comments about these.

Reviewer: Can we change the reference to 'model' to be more generic; for example "....between analysis" as I want to make sure this is accessible to people doing any analysis and not just modelling

Also, for the final sentence could you amend to (something like):  Shiny apps sit somewhere between analysis and an IT product and these are covered in more detail below.

There are often four types of software testing covered:

> * Unit testing (tests the different components or "units" of code using tools such as `testthat` in R).

Reviewer: For these bullets could we use a hyphen instead of brackets please, I think that's more consistent with the rest of the document

* Integration testing (tests the correct flow of data through the components).
* System testing (tests the overall outputs make sense).
* Acceptance testing (tests whether or not the work meets its objectives).

> Software testing can make a distinction between functional and non-functional testing too. In the context of an analytical model, a functional test would be one which tests the model produces what its intended to whereas a non-functional test would cover things such as its run-time and interface (i.e. things which affect the experience of using it, but not the numbers that come out).

Reviewer: Can you replace 'analytical model' with 'analysis' and 'tests the model' with 'tests that the analysis'

> For more information, see [here](https://www.guru99.com/levels-of-testing.html) or use your favourite search engine to explore the topic.

Reviewer: Can we bring this sentence up into the previous paragraph and amend is so that it's along the lines of "You can find more information here" so that it's consistent with the writing style of the whole document

Author: I've brought this into the paragraph but I've not added "you" in given I was on a spree of removing references to you. Having scanned other sections they don't seem to particularly use "you" when referring to other sources. Please let me know if you think I've missed something.

> As this is a chapter about testing code, we are only going to cover unit testing in more detail. The other testing aspects cover testing of wider areas of a software project than just the code, and other parts of the DfE QA framework cover how to handle these:

Reviewer: Can we frame the first sentence more positively?  Eg so that it's not "we are ONLY going to cover..." and more like "We are going to focus on unit testing in more detail as this will apply to most code"

* Integration testing tests correct data flow, transformations and interactions between different model components.
* System testing relates to verification tests on the overall outputs.
* Acceptance testing relates to the validation pillar of the QA framework.

### Unit testing

> Unit testing seeks to test every distinct component within your code to enable to you isolate specific areas that may be introducing errors. If you have written you code using functions, it is sometimes easier to think of this as testing each specific function.

Reviewer: Can we remove references to you/your please?  I've just skimmed through the guide and it doesn't use you/your very much so I think we need to keep this as consistent as possible.  Perhaps "....component within the code to enable an analyst to isolate...... If code has been written using functions, it is sometimes...." but open to suggestions here as that's probably a little clunky still
We should also link to the 'Optimising Code Structure' chapter as well in the word 'functions'

Author: I have linked to optimising code structure now, but where I refer to different code files not functions given it talks about different files and not about functions.

One of the key principles of unit testing is that each unit (or function) should be tested independently of all the others and without dependency on any data sources that may change. The objective is to isolate the cause of any errors:

> * If you test a function relying on output from previous functions and that test throws an error, you do not immediately know which of the functions (or units) caused the error and therefore it is not a unit test.

Reviewer: As line 60, this needs rephrasing to remove the you/your references, eg: "If a test relies on an output from a previous function and it throes an error, it is not obvious which of the functions caused the error

> * If you test a function relying on data from a live source (e.g SQL), it is possible your data could change and cause you test to fail without your function (unit) having any issues so this is also not a unit test.

Reviewer: Typo:  you instead of your

> * If you test a function relying on data from a live source (e.g SQL), it is possible your data could change and cause you test to fail without your function (unit) having any issues so this is also not a unit test.

Reviewer: As above; something like "Tests that rely on data from live sources (eg SQL) could fail because of a change in the data, as opposed to your function/unit, so this is also not a unit test

> To write a good set of unit tests, you need to think about the different types of input (e.g. positive values, zero values) your function may receive, define some fixed inputs to cover these and the write your tests to ensure the different cases are handled correctly.

Reviewer: As above, this needs rephrasing to remove 'you'

## Formalised testing

> Beyond in-script tests, an analyst may decide to take a more formal approach to testing if the project warrants it. Software development has dedicated testing professionals (called "testers") and there is plenty of literature about software testing on the internet. Much of this is transferable to code based analysis, but not all of it. There are times when it makes sense to make a distinction between a piece of analysis (which is generally operated by a technical specialist) and an IT product (e.g. a website, which is used by a broader user base). Shiny apps sit somewhere between analysis and an IT product so we'll include some specific comments about these.

Reviewer: ```suggestion
Beyond in-script tests, an analyst may decide to take a more formal approach to testing if the project warrants it. Software development has dedicated testing professionals (called "testers") and there is plenty of literature about software testing on the internet. Much of this is transferable to code based analysis, but not all of it. There are times when it makes sense to make a distinction between a piece of analysis (which is generally operated by a technical specialist) and an IT product (e.g. a website, which is used by a broader user base). Shiny apps sit somewhere between analysis and an IT product therefore these are covered separately later on.
```

> There are often four types of software testing covered:

Reviewer: ```suggestion
The following types of software testing are commonly used:
```

* Unit testing - tests the different components or "units" of code using tools such as `testthat` in R.
* Integration testing - tests the correct flow of data through the components.
* System testing - tests the overall outputs make sense.
* Acceptance testing - tests whether or not the work meets its objectives.

> Software testing can make a distinction between functional and non-functional testing too. In the context of analysis, a functional test would be one which tests the analysis produces what its intended to whereas a non-functional test would cover things such as its run-time and interface (i.e. things which affect the experience of using it, but not the numbers that come out). For more information, see [here](https://www.guru99.com/levels-of-testing.html) or explore further with help from a search engine.

Reviewer: Nicky's comment on this section was:  Given that the focus of the remainder of the chapter is on unit testing I think the flow would be better to talk about unit testing and other testing, bundling integration, system and acceptance as they are more about other aspects, as Rachel states in the text.

So, my suggestion is to drop this paragraph but add in the 'For more information.... sentence at the end of the section where you say more about what the other tests relate to but if you think you can address that comment in another way, please go ahead!

Author: I have added in another section after the "unit testing" section for some of this content. Having thought a bit about what Nicky was saying, I think this matches up. Also, having made the change I think I prefer it.

> As this is a chapter about testing code, we going to focus primarily on unit testing. The other testing aspects cover testing of wider areas of a software project than just the code, and other parts of the DfE QA framework cover how to handle these:

Reviewer: typo:  missing 'are' from we are going to....

> As this is a chapter about testing code, we going to focus primarily on unit testing. The other testing aspects cover testing of wider areas of a software project than just the code, and other parts of the DfE QA framework cover how to handle these:

Reviewer: ```suggestion
This chapter focuses on unit testing as the other testing aspects cover testing of wider areas of a software project than just the code, and other parts of the DfE QA framework cover how to handle these:
```

* Integration testing tests correct data flow, transformations and interactions between different parts of the analysis.
* System testing relates to verification tests on the overall outputs.

> * Acceptance testing relates to the validation pillar of the QA framework.

Reviewer: ```suggestion
* Acceptance testing relates to the validation pillar of the QA framework.

For more information, see [here](https://www.guru99.com/levels-of-testing.html) or explore further with help from a search engine.
```

### Unit testing

Unit testing seeks to test every distinct code component to enable isolation of specific areas that may be introducing errors. For code made up of functions, it is sometimes easier to think of this as testing each specific function.

> One of the key principles of unit testing is that each unit (or function) should be tested independently of all the others and without dependency on any data sources that may change. The objective is to isolate the cause of any errors:

Reviewer: ```suggestion
One of the key principles of unit testing is that each unit (or function) should be tested independently of all the others and without dependency on any data sources that may change. The objective is to isolate the cause of any errors which means that unit tests should not rely on output from previous functions or rely on data from a live source (e.g SQL) as in both these instances it would not be possible to identify if the function being tested caused the error.
```

> * Testing a function relying on output from previous functions means that, if the test throws an error, it is not immediately clear of the functions (or units) caused the error and therefore it is not a unit test.

Reviewer: typo:  clear of the functions should be: clear if the functions....

> * Testing a function relying on output from previous functions means that, if the test throws an error, it is not immediately clear of the functions (or units) caused the error and therefore it is not a unit test.

Reviewer: Need to remove this and the following bullet now

* Testing a function relying on data from a live source (e.g SQL) means it is possible the data could change and cause the test to fail without the function (unit) having any issues so this is also not a unit test.

> To write a good set of unit tests, think about the different types of input (e.g. positive values, zero values) the function may receive, define some fixed inputs to cover these and the write tests to ensure the different cases are handled correctly.

Reviewer: typo:  and then write (or just and write) but not and the write

> To write a good set of unit tests, think about the different types of input (e.g. positive values, zero values) the function may receive, define some fixed inputs to cover these and the write tests to ensure the different cases are handled correctly.

Reviewer: ```suggestion
To write a good set of unit tests, think about the different types of input (e.g. positive values, zero values) the function may receive, define some fixed inputs to cover these and then write tests to ensure the different cases are handled correctly.
```

> To implement unit tests, it is generally easiest to use a unit testing framework. These generally allow tests to be written separate files and then all tests run at once, returning success/failure feedback.

Reviewer: typo:  tests to be written IN separate files

The `testthat` package, maintained by the RStudio team, is commonly used for unit testing in R. Although most of its documentation and features focus on testing in the context of package writing, `testthat` can also be used with regular R projects.  There's a brief overview of unit testing below, but for a more detailed introduction see [here](https://katherinemwood.github.io/post/testthat/) or [here](https://www.r-bloggers.com/automated-testing-with-testthat-in-practice/) for a more in-depth overview of automated testing, including test driven development. 

> In `testthat`, tests are called **expectations**. These spell out the expected behaviour of the function or unit of code. `testthat` provides many functions to define expectations - see details [here](http://r-pkgs.had.co.nz/tests.html). It is good practice to save tests relating to the [different code files](#code_structure) in different test files, all saved within a dedicated testing directory. Below is an example of testing a simple function.

Reviewer: ```suggestion
In `testthat`, tests are called **expectations**. These spell out the expected behaviour of the function or unit of code. `testthat` provides many functions to define expectations - see details [here](http://r-pkgs.had.co.nz/tests.html). Expectations falling within a same testing context are stored in a same file and all testing files are saved within a dedicated testing directory.  Below is an example of testing a simple function.
```

> Firstly we need to have a defined function to test:

Reviewer: There's a typo lower down that I can't tag where you've said:  Then we can defined ... so the d needs dropping so it's just define

# The second test checks that the maximum value
# of the variable "some_numbers" does not exceed 1
test_that('no value greater than 1', {
  expect_lte(max(mydata$some_numbers), 1)
})

```

Expectations are tested using the `test_dir()` function, which takes the path to the testing directory as an argument:
```{r testing, message = FALSE, eval = FALSE, class.source = "badCode"}
# test expectations and examine outputs
# In this case, all tests are OK
test_dir("testing_directory")
```

`test_dir()` provides detailed outputs, including a time line of success/failure and any warnings that may have occurred.

### Testing Shiny apps

> R Shiny apps may benefit from a further extended testing approach. They encompass not just some analysis, but also a User Interface and a Server to enable them to run. This makes them closer to a software development project than the average piece of modeling work.

Reviewer: '"..development project than a piece of analysis'

Additional functionality is available within R to support testing of Shiny apps. Further details about this can be found [here](https://shiny.rstudio.com/articles/integration-testing.html).

### Test driven development

> "Test driven development" describes a way of working in software where tests are written prior to the code being written. This forces developers to sit down and work out what working code encompasses before the write it and avoids any retrofitting of tests to match achieved output. If you have a strong QA plan, you could also implement this for analysis. As testing of each new bit of code happens immediately after creation it also reduced the risk of something faulty getting heavily baked in.

Reviewer: Do we need 'in software' ?  If we want people to adopt this approach then I think removing that could be beneficial so it doesn't seem like it's only applicable in software dev.

On the 2nd sentence, could we rephrase more positively?  Eg:  "By thinking about what the working code will need to encompass before its written, this helps to avoid retrofitting tests to match achieved output"

Also, typo on 'it also reduceS' (not reduced)

### Testing Shiny apps

> R Shiny apps may benefit from a further extended testing approach. They encompass not just some analysis, but also a User Interface and a Server to enable them to run. This makes them closer to a software development project than the average piece of analysis work. Additional functionality is available within R to support testing of Shiny apps. Further details about this can be found [here](https://shiny.rstudio.com/articles/integration-testing.html).

Reviewer: ```suggestion
R Shiny apps may benefit from a further extended testing approach. They encompass not just some analysis, but also a User Interface and a Server to enable them to run. This makes them closer to a software development project than the average piece of analytical work. Additional functionality is available within R to support testing of Shiny apps. Further details about this can be found [here](https://shiny.rstudio.com/articles/integration-testing.html).
```
